sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -83,7 +83,25 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
         numPartitionsSet.headOption
       }
 
-      val targetNumPartitions = requiredNumPartitions.getOrElse(childrenNumPartitions.max)
+      // maxNumPostShufflePartitions is usually larger than numShufflePartitions,
+      // which causes some bucket map join lose efficacy after enabling adaptive execution.
+      // Please see SPARK-29655 for more details.
+      val expectedChildrenNumPartitions = if (conf.adaptiveExecutionEnabled) {
```

Comment:
The logical is convoluted here as we've already added the shuffles with `maxNumPostShufflePartitions`, and we need to revert it.

Can we make the implementation clearer? Basically we are picking the `targetNumPartitions` as:
1. if there are no non-shuffle children, keep the previous behavior.
2. for non-shuffle children, get the max num partitions among them.
  2.1. if the max num partitions is larger than `conf.numShufflePartitions`, pick it as `targetNumPartitions`.
  2.2. otherwise, pick `conf.numShufflePartitions` as `targetNumPartitions`

Reply:
Yes. This implementation could make it clearer.

Reply:
What do you think of another implementation: https://github.com/apache/spark/pull/26409/commits/45109c71c3efc31f0d5aba6588a811faeea2a8f2? It may be more reasonable. But personally, I do not like add new conf.

Reply:
it's indeed clearer to have a dedicated config, but I'd like to have a better solution that can make decisions automatically instead of relying on configs. We can think about it later.